batchproducer/batchproducer.go

```diff
@@ -233,15 +232,22 @@ func (b *batchProducer) Stop() error {
 }
 
 func (b *batchProducer) setRunning(running bool) {
-	b.runningMu.Lock()
-	defer b.runningMu.Unlock()
-	b.running = running
+	var newValue int32
+	if running {
+		newValue = 1
+	} else {
+		newValue = 0
+	}
+
+	oldValue := b.running
+
+	for swapped := false; !swapped; {
```

Comment:
Two more thoughts:

- Very pedantic, but we can have `b.stop` be a `chan struct{}` instead of a `chan bool` (the whole `struct{}` takes 0 bytes deal)
- We can introduce a new `b.done`, to which we push a `struct{}{}` on a `defer` at `run()`:

```go
func (...) run() {
  // do stuff

  defer b.done<- struct{}{}

  for {
    // loop until empty
  }
}
```

And then on `Stop()` (or `StopAndWait()`) we can do this:

func (b *batchProducer) Stop() error {
  if !atomic.CompareAndSwapInt32(&b.running, 1, 0) {
    return ErrAlreadyStopped
  }
  // at this point b.isRunning() will return `false`
  b.stop<- true

  // wait for queue to drain
  <-b.done
  return nil
}